### PR TITLE
[Opset13] ScaledDotProductAttention-13 input validation improvements

### PR DESCRIPTION
**Pull request by mmikolajcz**
### Details:
 - *Improve ScaledDotProductAttention shape validation/inference for inputs and outputs*
 - *Improve type validation and inference*
 - *Update type_prop tests*

### Tickets:
 - *126323*

### CODE REVIEW

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            OPENVINO_ASSERT(Dimension::merge(l_dim, l_dim, *(attention_mask.end() - 2)));
+            OPENVINO_ASSERT(Dimension::merge(s_dim, s_dim, *(attention_mask.end() - 1)));
+            OPENVINO_ASSERT(TRShape::broadcast_merge_into(m_dims, n_dims, AutoBroadcastType::NUMPY));
+            OPENVINO_ASSERT(n_dims.compatible(m_dims));
+        }
+    }
+
+    if (has_scale_input) {
+        const auto scale = input_shapes[4];
+        OPENVINO_ASSERT(scale.rank().is_dynamic() || scale.rank() == 0, "Scale input accepts only scalar tensor.");
```
Comment (ilya-lavrenov) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:65]: too much exceptions for 70 lines of code..

Comment (mmikolajcz) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:65]: Yeah, I agree, it's a bit tricky op to validate everything. For now it's just initial draft, will look for way to simplify this tree of asserts

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    TRShape n_dims{TRShape::dynamic()};
+    Dimension s_dim{Dimension::dynamic()};
+    Dimension e_dim{Dimension::dynamic()};
+    Dimension l_dim{Dimension::dynamic()};
+    Dimension ev_dim{Dimension::dynamic()};
+    TRShape m_dims{TRShape::dynamic()};
+
+    const auto query = input_shapes[0];
+    if (query.rank().is_static()) {
+        OPENVINO_ASSERT(query.rank().get_length() >= 3);
```
Comment (mitruska) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:27]: Please use `NODE_SHAPE_INFER_CHECK` instead of `OPENVINO_ASSERT`,
and assign result of several conditions of Dimension::merge to a bool variable, then reuse the variable in the check.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+namespace op {
+namespace v13 {
+template <class T, class TRShape = result_shape_t<T>>
+std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op, const std::vector<T>& input_shapes) {
+    const auto inputs_count = input_shapes.size();
+    const auto has_attention_mask_input = inputs_count >= 4;
+    const auto has_scale_input = inputs_count == 5;
+    NODE_VALIDATION_CHECK(op, inputs_count == 3 || has_attention_mask_input || has_scale_input);
+    TRShape n_dims{TRShape::dynamic()};
+    Dimension s_dim{Dimension::dynamic()};
```
Comment (PiotrKrzem) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:19]: Suggestion to be investigated - instead of using single dim type `Dimension`, the following is commonly used instead:
`using DimType = typename T::value_type;`

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp (hunk 2 above)
Comment (mmikolajcz) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: Done

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp (hunk 3 above)
Comment (mmikolajcz) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: Done

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+#include "openvino/op/scaled_dot_product_attention.hpp"
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace v13 {
+template <class T, class TRShape = result_shape_t<T>>
+std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
+                                 const std::vector<T>& input_shapes,
+                                 const bool& iscausal) {
```
Comment (praasz) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:15]:
```suggestion
                                 const bool& iscausal) {
```
Remove this parameter, it will not fit for common signature of this function

**Review by praasz — CHANGES_REQUESTED** [review]

File: src/core/src/op/scaled_dot_product_attention.cpp [hunk 5]
```diff
@@ -64,7 +64,7 @@ void op::v13::ScaledDotProductAttention::validate_and_infer_types() {
                           "The element type of the input tensor must be a floating-point.");
 
     OPENVINO_SUPPRESS_DEPRECATED_START
-    const auto input_shapes = get_node_input_partial_shapes(*this);
+    const auto input_shapes = ov::util::get_node_input_partial_shapes(*this);
     OPENVINO_SUPPRESS_DEPRECATED_END
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/src/op/scaled_dot_product_attention.cpp]: Done

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp (hunk 4 above)
Comment (mmikolajcz) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: Done

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                                   attention_mask_input_correctness,
+                                   "Attention mask input shape not compatible with other inputs.");
+        }
+    }
+
+    if (has_scale_input) {
+        const auto scale = input_shapes[4];
+        if (scale.rank().is_static()) {
+            NODE_SHAPE_INFER_CHECK(op, input_shapes, scale.rank() == 0, "Scale input accepts only scalar tensor.");
+        }
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: Done

File: src/core/src/op/scaled_dot_product_attention.cpp [hunk 7]
```diff
@@ -38,25 +39,35 @@ op::v13::ScaledDotProductAttention::ScaledDotProductAttention(const Output<Node>
...
+    const auto causal = get_causal();
+    if (input_size >= 4 && !causal) {
+        auto attention_type = get_input_element_type(3);
+        NODE_VALIDATION_CHECK(
+            this,
+            attention_type.is_real() || attention_type == element::boolean || attention_type.is_dynamic(),
+            "The element type of attention_mask must be either floating-point or boolean.");
     }
+    for (size_t i = 1; i < input_size; i++) {
+        auto element_type = get_input_element_type(i);
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/src/op/scaled_dot_product_attention.cpp]: Done

File: src/core/src/op/scaled_dot_product_attention.cpp [hunk 8]
```diff
@@ -38,25 +39,35 @@ op::v13::ScaledDotProductAttention::ScaledDotProductAttention(const Output<Node>
...
-
-    // using particular dimensions from query and value, to do that need to have them statically ranked
-    if (query.rank().is_dynamic() || value.rank().is_dynamic()) {
-        set_output_type(0, get_input_element_type(0), PartialShape::dynamic());
-        return;
+    auto out_type = get_input_element_type(0);
+    const auto input_size = get_input_size();
+    const auto causal = get_causal();
+    if (input_size >= 4 && !causal) {
+        auto attention_type = get_input_element_type(3);
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/src/op/scaled_dot_product_attention.cpp]: Done

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
+                                 const std::vector<T>& input_shapes,
+                                 const bool& iscausal) {
+    using DimType = typename T::value_type;
+    const auto inputs_count = input_shapes.size();
+    const auto has_attention_mask_input = inputs_count >= 4;
+    const auto has_scale_input = inputs_count == 5;
+    NODE_VALIDATION_CHECK(op, inputs_count == 3 || has_attention_mask_input || has_scale_input);
+    T n_dims{T::dynamic()};
+    DimType s_dim{DimType::dynamic()};
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp]: Thanks, added static dimension test in CPU plugin + modified shape infer to accept it

File: src/core/tests/type_prop/scaled_dot_product_attention.cpp [hunk 10]
```diff
@@ -12,156 +12,408 @@
...
+    EXPECT_EQ(op->get_output_element_type(0), element::f64);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{2, 3, 6}));
+}
+
+TEST(type_prop, scaled_dot_product_attention_static_5_inputs_broadcast) {
+    const auto query = std::make_shared<opset13::Parameter>(element::f32, PartialShape{2, 1, 3, 1});
+    const auto key = std::make_shared<opset13::Parameter>(element::f32, PartialShape{1, 7, 1, 4});
+    const auto value = std::make_shared<opset13::Parameter>(element::f32, PartialShape{2, 7, 5, 1});
+    const auto attention_mask = std::make_shared<opset13::Parameter>(element::f32, PartialShape{1, 1, 1, 5});
+    const auto scale = std::make_shared<opset13::Parameter>(element::f32, PartialShape{1});
```
Comment (PiotrKrzem) [inline src/core/tests/type_prop/scaled_dot_product_attention.cpp:304]: SHould we be testing those if they do not match the specifications of the op?

Comment (mmikolajcz) [reply src/core/tests/type_prop/scaled_dot_product_attention.cpp]: It seem that batch broadcasts are required by models according to https://github.com/openvinotoolkit/openvino/pull/21217#pullrequestreview-1747252695. There were also 1 additional broadcast for L shape in CPU test for L dim, however, with it disabled, models still seem to pass, so I've modified CPU test to remove it

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 11]
```diff
@@ -88,9 +92,10 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
     }
 
     if (has_scale_input) {
-        if (input_shapes[4].rank().is_static() && input_shapes[4].is_static()) {
-            const auto& scale_is_scalar = input_shapes[4].rank().compatible(0);
-            const auto& scale_has_one_elem = input_shapes[4].rank().compatible(1) && input_shapes[4][0].compatible(1);
+        const auto& scale_rank = input_shapes[4].rank();
+        if (scale_rank.is_static() || input_shapes[4].is_static()) {
```
Comment (praasz) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:96]: I think only check the rank is enough.

File: src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp [hunk 12]
```diff
@@ -88,9 +92,10 @@ std::vector<TRShape> shape_infer(const ScaledDotProductAttention* op,
...
+            const auto& scale_is_scalar = scale_rank.compatible(0);
+            const auto& scale_has_one_elem = scale_rank.compatible(1) && input_shapes[4][0].compatible(1);
```
Comment (praasz) [inline src/core/shape_inference/include/scaled_dot_product_attention_shape_inference.hpp:98]: This is not exactly valid as compatible return true if rank is dynamic so accessing the dimension[0] is not valid then.

But as main check checks rank is static the rank length can be taken and checked if is 0 or 1 etc.

**Review by praasz — APPROVED** [review]
